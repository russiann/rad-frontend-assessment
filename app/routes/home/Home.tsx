import React from "react";
import type { Route } from "./+types/Home";
import { useHome } from "./useHome";
import { FilterSection } from "./partials/FilterSection";
import { ProductGrid } from "./partials/ProductGrid";

export function meta({}: Route.MetaArgs) {
  return [
    { title: "Products - React Shopping" },
    { name: "description", content: "Browse our amazing products!" },
  ];
}

export default function Home() {
  const { state, actions } = useHome();

  return (
    <div className="mb-8">
      <div className="flex flex-col gap-6">
        <div>
          <FilterSection
            selectedCategory={state.selectedCategory}
            selectedSort={state.selectedSort}
            onCategoryFilter={actions.filterByCategory}
            onSortChange={actions.sortProducts}
          />
        </div>
        
        <div>
          <ProductGrid
            products={state.filteredProducts}
            onAddToCart={actions.addToCart}
          />
        </div>
      </div>
    </div>
  );
} 