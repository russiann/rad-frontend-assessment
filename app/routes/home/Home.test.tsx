import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import Home from './Home';
import type { Product } from '../../utils/trpc';
import type { ReactNode } from 'react';

// Mock dependencies
const mockAddToCart = vi.fn();
const mockOnOpenChange = vi.fn();

vi.mock('../../contexts/CartContext', () => ({
  useCartContext: () => ({
    actions: {
      addToCart: mockAddToCart,
    },
  }),
}));

vi.mock('../../contexts/ChatContext', () => ({
  useChatContext: () => ({
    isOpen: false,
    onOpenChange: mockOnOpenChange,
  }),
}));

vi.mock('../../utils/notifications', () => ({
  showPriceUpdateToast: vi.fn(),
  showAddedToCartToast: vi.fn(),
}));

// Mock react-router
vi.mock('react-router', () => ({
  Link: ({ children, to }: { children: React.ReactNode; to: string }) => (
    <a href={to}>{children}</a>
  ),
  useLocation: () => ({ pathname: '/' }),
  useParams: () => ({}),
}));

// Mock products
const mockProducts: Product[] = [
  {
    id: 1,
    name: 'Gaming Laptop',
    price: 999.99,
    category: 'Electronics',
    description: 'High-performance gaming laptop',
    image: '/laptop.jpg',
    isAvailable: true,
    stock: 10,
    createdAt: '2024-01-01T00:00:00Z',
    updatedAt: '2024-01-01T00:00:00Z',
  },
  {
    id: 2,
    name: 'Cotton T-Shirt',
    price: 29.99,
    category: 'Clothing',
    description: 'Comfortable cotton t-shirt',
    image: '/tshirt.jpg',
    isAvailable: true,
    stock: 5,
    createdAt: '2024-01-01T00:00:00Z',
    updatedAt: '2024-01-01T00:00:00Z',
  },
  {
    id: 3,
    name: 'Running Shoes',
    price: 79.99,
    category: 'Footwear',
    description: 'Lightweight running shoes',
    image: '/shoes.jpg',
    isAvailable: false,
    stock: 0,
    createdAt: '2024-01-01T00:00:00Z',
    updatedAt: '2024-01-01T00:00:00Z',
  },
];

const mockUseQuery = vi.fn();

vi.mock('../../utils/trpc', () => ({
  trpc: {
    product: {
      list: {
        useQuery: () => mockUseQuery(),
      },
    },
  },
}));

describe('Home Page', () => {
  let queryClient: QueryClient;
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
        },
      },
    });

    user = userEvent.setup();
    mockAddToCart.mockClear();
    mockUseQuery.mockReturnValue({
      data: mockProducts,
      isLoading: false,
    });
  });

  afterEach(() => {
    queryClient.clear();
  });

  const createWrapper = () => ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  describe('Basic Rendering', () => {
    it('should render filter section and product grid', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      expect(screen.getByText('Filter by Category')).toBeInTheDocument();
      expect(screen.getByText('Sort by Price')).toBeInTheDocument();
      expect(screen.getByText('Gaming Laptop')).toBeInTheDocument();
      expect(screen.getByText('Cotton T-Shirt')).toBeInTheDocument();
      expect(screen.getByText('Running Shoes')).toBeInTheDocument();
    });

    it('should display product prices correctly', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      expect(screen.getByText('$999.99')).toBeInTheDocument();
      expect(screen.getByText('$29.99')).toBeInTheDocument();
      expect(screen.getByText('$79.99')).toBeInTheDocument();
    });

    it('should show unavailable products with correct styling', () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      expect(screen.getAllByText('Unavailable')).toHaveLength(2); // Badge and button
      const unavailableButton = screen.getByRole('button', { name: /unavailable/i });
      expect(unavailableButton).toBeDisabled();
    });
  });

  describe('Category Filtering', () => {
    it('should filter products by Electronics category', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      const categoryButton = screen.getByText('Filter by Category');
      await user.click(categoryButton);

      await waitFor(() => {
        const dropdownOptions = screen.getAllByText('Electronics');
        expect(dropdownOptions.length).toBeGreaterThan(0);
      });

      // Click on the dropdown option (not the product category)
      const dropdownOptions = screen.getAllByText('Electronics');
      const dropdownOption = dropdownOptions.find(option => 
        option.closest('[role="option"]') || option.closest('[data-key="electronics"]')
      ) || dropdownOptions[dropdownOptions.length - 1]; // fallback to last option
      
      await user.click(dropdownOption);

      // Should only show electronics products
      await waitFor(() => {
        expect(screen.getByText('Gaming Laptop')).toBeInTheDocument();
        expect(screen.queryByText('Cotton T-Shirt')).not.toBeInTheDocument();
        expect(screen.queryByText('Running Shoes')).not.toBeInTheDocument();
      });
    });

    it('should show all products when "All Categories" is selected', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      // First filter by Electronics
      const categoryButton = screen.getByText('Filter by Category');
      await user.click(categoryButton);
      
      await waitFor(() => {
        expect(screen.getByText('All Categories')).toBeInTheDocument();
      });
      
      await user.click(screen.getByText('All Categories'));

      await waitFor(() => {
        expect(screen.getByText('Gaming Laptop')).toBeInTheDocument();
        expect(screen.getByText('Cotton T-Shirt')).toBeInTheDocument();
        expect(screen.getByText('Running Shoes')).toBeInTheDocument();
      });
    });
  });

  describe('Price Sorting', () => {
    it('should sort products by price low to high', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      const sortButton = screen.getByText('Sort by Price');
      await user.click(sortButton);

      await waitFor(() => {
        expect(screen.getByText('Low to High')).toBeInTheDocument();
      });

      await user.click(screen.getByText('Low to High'));

      await waitFor(() => {
        const productCards = screen.getAllByText(/\$\d+\.\d+/);
        // First price should be $29.99 (Cotton T-Shirt)
        expect(productCards[0]).toHaveTextContent('$29.99');
      });
    });

    it('should sort products by price high to low', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      const sortButton = screen.getByText('Sort by Price');
      await user.click(sortButton);
      await user.click(screen.getByText('High to Low'));

      await waitFor(() => {
        const productCards = screen.getAllByText(/\$\d+\.\d+/);
        // First price should be $999.99 (Gaming Laptop)
        expect(productCards[0]).toHaveTextContent('$999.99');
      });
    });
  });

  describe('Add to Cart Functionality', () => {
    it('should add available product to cart', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      const addToCartButtons = screen.getAllByText('Add to Cart');
      await user.click(addToCartButtons[0]);

      expect(mockAddToCart).toHaveBeenCalledWith(mockProducts[0]);
    });

    it('should not add unavailable product to cart', async () => {
      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      const unavailableButton = screen.getByRole('button', { name: /unavailable/i });
      expect(unavailableButton).toBeDisabled();
      
      // Try to click but it shouldn't work since it's disabled
      await user.click(unavailableButton);
      expect(mockAddToCart).not.toHaveBeenCalled();
    });
  });

  describe('Loading State', () => {
    it('should handle loading state gracefully', () => {
      mockUseQuery.mockReturnValue({
        data: [],
        isLoading: true,
      });

      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      // Should still render filter section even during loading
      expect(screen.getByText('Filter by Category')).toBeInTheDocument();
      expect(screen.getByText('Sort by Price')).toBeInTheDocument();
    });
  });

  describe('Combined Operations', () => {
    it('should apply both filtering and sorting together', async () => {
      // Add another electronics product to make sorting more meaningful
      const extendedProducts = [
        ...mockProducts,
        {
          id: 4,
          name: 'Smartphone',
          price: 599.99,
          category: 'Electronics',
          description: 'Latest smartphone',
          image: '/phone.jpg',
          isAvailable: true,
          stock: 10,
          createdAt: '2024-01-01T00:00:00Z',
          updatedAt: '2024-01-01T00:00:00Z',
        },
      ];

      mockUseQuery.mockReturnValue({
        data: extendedProducts,
        isLoading: false,
      });

      const TestWrapper = createWrapper();
      
      render(
        <TestWrapper>
          <Home />
        </TestWrapper>
      );

      // Filter by Electronics
      const categoryButton = screen.getByText('Filter by Category');
      await user.click(categoryButton);
      
      await waitFor(() => {
        const dropdownOptions = screen.getAllByText('Electronics');
        expect(dropdownOptions.length).toBeGreaterThan(0);
      });

      // Click on the dropdown option
      const dropdownOptions = screen.getAllByText('Electronics');
      const dropdownOption = dropdownOptions.find(option => 
        option.closest('[role="option"]') || option.closest('[data-key="electronics"]')
      ) || dropdownOptions[dropdownOptions.length - 1];
      
      await user.click(dropdownOption);

      // Sort by low to high
      const sortButton = screen.getByText('Sort by Price');
      await user.click(sortButton);
      
      await waitFor(() => {
        expect(screen.getByText('Low to High')).toBeInTheDocument();
      });
      
      await user.click(screen.getByText('Low to High'));

      await waitFor(() => {
        // Should show only electronics products, sorted by price
        expect(screen.getByText('Smartphone')).toBeInTheDocument();
        expect(screen.getByText('Gaming Laptop')).toBeInTheDocument();
        expect(screen.queryByText('Cotton T-Shirt')).not.toBeInTheDocument();
      });
    });
  });

  // Snapshots temporarily disabled - will re-enable after fixing component rendering
  // describe('snapshots', () => {
  //   it('renders home page correctly', () => {
  //     const TestWrapper = createWrapper();
      
  //     const { container } = render(
  //       <TestWrapper>
  //         <Home />
  //       </TestWrapper>
  //     );

  //     expect(container.firstChild).toMatchSnapshot();
  //   });
  // });
});