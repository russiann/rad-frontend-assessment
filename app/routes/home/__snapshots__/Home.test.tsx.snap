// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Page > snapshots > renders home page correctly 1`] = `
<div
  class="mb-8"
>
  <div
    class="flex flex-col gap-6"
  >
    <div>
      <div
        class="flex flex-col sm:flex-row gap-4 mb-6"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
          data-react-aria-pressable="true"
          data-slot="trigger"
          id="react-aria-«r68»"
          tabindex="0"
          type="button"
        >
          Filter by Category
          <svg
            aria-hidden="true"
            class="iconify iconify--lucide text-sm"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <path
              d="m6 9l6 6l6-6"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
          data-react-aria-pressable="true"
          data-slot="trigger"
          id="react-aria-«r6b»"
          tabindex="0"
          type="button"
        >
          Sort by Price
          <svg
            aria-hidden="true"
            class="iconify iconify--lucide text-sm"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <path
              d="m6 9l6 6l6-6"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
    <div>
      <div
        class="grid gap-6"
        style="grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));"
      >
        <div
          class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
          tabindex="-1"
        >
          <a
            href="/product/1"
          >
            <div
              class="relative "
            >
              <img
                alt="Gaming Laptop"
                class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
                src="/laptop.jpg"
              />
            </div>
          </a>
          <div
            class="p-4 space-y-3"
          >
            <a
              href="/product/1"
            >
              <h3
                class="font-semibold text-lg hover:text-primary transition-colors "
              >
                Gaming Laptop
              </h3>
            </a>
            <div
              class="flex items-center justify-between"
            >
              <p
                class="font-medium text-default-600"
              >
                $
                999.99
              </p>
              <p
                class="text-xs text-default-500"
              >
                Electronics
              </p>
            </div>
            <button
              class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full"
              data-react-aria-pressable="true"
              tabindex="0"
              type="button"
            >
              Add to Cart
            </button>
          </div>
        </div>
        <div
          class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
          tabindex="-1"
        >
          <a
            href="/product/2"
          >
            <div
              class="relative "
            >
              <img
                alt="Cotton T-Shirt"
                class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
                src="/tshirt.jpg"
              />
            </div>
          </a>
          <div
            class="p-4 space-y-3"
          >
            <a
              href="/product/2"
            >
              <h3
                class="font-semibold text-lg hover:text-primary transition-colors "
              >
                Cotton T-Shirt
              </h3>
            </a>
            <div
              class="flex items-center justify-between"
            >
              <p
                class="font-medium text-default-600"
              >
                $
                29.99
              </p>
              <p
                class="text-xs text-default-500"
              >
                Clothing
              </p>
            </div>
            <button
              class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full"
              data-react-aria-pressable="true"
              tabindex="0"
              type="button"
            >
              Add to Cart
            </button>
          </div>
        </div>
        <div
          class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
          tabindex="-1"
        >
          <div
            class="relative inline-flex shrink-0"
          >
            <span />
            <span
              class="flex flex-wrap absolute box-border rounded-full whitespace-nowrap place-content-center origin-center items-center select-none font-regular scale-100 opacity-100 subpixel-antialiased data-[invisible=true]:scale-0 data-[invisible=true]:opacity-0 px-1 text-small transition-transform-opacity !ease-soft-spring !duration-300 border-2 border-background bg-danger text-danger-foreground top-[5%] right-[5%] translate-x-1/2 -translate-y-1/2 z-10 m-2 cursor-pointer"
            >
              SALE
            </span>
          </div>
          <div
            class="max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-6 text-tiny rounded-full bg-danger/20 text-danger-600 dark:text-danger-500 absolute top-2 left-2 z-10"
          >
            <span
              class="flex-1 text-inherit font-normal px-1"
            >
              Unavailable
            </span>
          </div>
          <a
            href="/product/3"
          >
            <div
              class="relative opacity-75"
            >
              <img
                alt="Running Shoes"
                class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
                src="/shoes.jpg"
              />
              <div
                class="absolute inset-0 bg-black bg-opacity-30 flex items-center justify-center"
              >
                <svg
                  aria-hidden="true"
                  class="iconify iconify--lucide text-white text-4xl"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                >
                  <g
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="m15 9l-6 6m0-6l6 6"
                    />
                  </g>
                </svg>
              </div>
            </div>
          </a>
          <div
            class="p-4 space-y-3"
          >
            <a
              href="/product/3"
            >
              <h3
                class="font-semibold text-lg hover:text-primary transition-colors text-default-400"
              >
                Running Shoes
              </h3>
            </a>
            <div
              class="flex items-center justify-between"
            >
              <p
                class="font-medium text-default-400"
              >
                $
                79.99
              </p>
              <p
                class="text-xs text-default-500"
              >
                Footwear
              </p>
            </div>
            <button
              class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small pointer-events-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full opacity-50 cursor-not-allowed"
              data-disabled="true"
              data-react-aria-pressable="true"
              disabled=""
              type="button"
            >
              Unavailable
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
